### PR TITLE
[DOP-9787] Improve read strategies in DBReader

### PR DESCRIPTION
**Pull request by dolfinus**
## Change Summary

1. Previously IncrementalStrategy was implemented like this:

- Get table schema by making query `SELECT * FROM table WHERE 1=0` (if `DBReader.columns` contains `*`)
- Append HWM column to list of table columns and remove duplicated columns.
- Create dataframe from query like `SELECT hwm.expression AS hwm.column, ...other table columns... FROM table WHERE prev_hwm.expression > prev_hwm.value`.
- Determine HWM class by `df.schema[hwm.column].dataType`
- Calculate `df.select(min(hwm.column), max(hwm.column)).collect()` on Spark side.
- Use `max(hwm.column)` as next HWM value.
- Return dataframe to user.

This was far from ideal:
- Dataframe content (all rows or just changed ones) was loaded from the source to Spark only to get min/max values of specific column.
- Step of fetching table schema and then substituting column names in the following query may cause errors.

  For example, source contains columns with mixed name case, like `"MyColumn"` and `"My column"`.
  Column names were not escaped during query generation, leading to queries that cannot be executed by database.
  So users have to explicitly set proper columns list with wrapping them with `"`.

- Dataframe was created from query with clause like `WHERE hwm.expression > prev_hwm.value`,
  not `WHERE hwm.expression > prev_hwm.value AND hwm.expression <= current_hwm.value`.

  So if new rows appeared in the source after HWM value is determined, these rows may be read by DBReader on the first run,
  and then again on the next run, because they are returned by `WHERE hwm.expression > prev_hwm.value` query.

Now it looks like this:
- Get type of HWM expression: `SELECT hwm.expression FROM table WHERE 1=0`
- Determine HWM class by `df.schema[0]`.
- Get min/max values by querying `SELECT MIN(hwm.expression), MAX(hwm.expression) FROM table WHERE hwm.expression >= prev_hwm.value`.
- Use `max(hwm.column)` as next HWM value.
- Create dataframe from query `SELECT * FROM table WHERE hwm.expression > prev_hwm.value AND hwm.expression <= current_hwm.value`, and return it to user.

Improvements:
- Allow source to calculate min/max instead of loading everything to Spark. This should be *really* fast, and also source can use indexes to speed this up even more.
- Restrict dataframe content to always match HWM values.
- Don't mess up with columns list, just pass them to source as-is. So `DBReader` does not fail on tables with mixed column naming.

**Breaking change** - HWM column is not being implicitly added to dataframe.
If it was not just some column value but some expression which then used in your code by accessing dataframe column,
you should explicitly add same expression to `DBReader.columns`.

2. Add 2 internal classes `Edge` and `Window`, replacing old `Statement` class. They represent `column > edge` / `column >= edge` / `column <= edge` parts of WHERE clause. Merged separated arguments `start_from`/`end_at` with `window` in `DBConnection` methods, added `apply_window` method to `DBDialect` which converts these classes to WHERE clause (replaces `condition_assembler`).

3. Get rid of `StrategyHelper` class, move corresponding logic to `DBReader` itself. Also update `Strategy` classes, properties `current`/`next` return `Edge` which simplifies implementation and allows to drop `current_comparator`/`next_comparator`.

4. Improved checks for different edge cases

- If HWMStore returned HWM of wrong type, an exception is raised.
- If multiple DBReaders with different HWMs are used within the same HWMStrategy, an exception is being raised.
- If HWMStore returned HWM with different entity or expression, a warning is shown and `DBReader.hwm` has higher priority than attributes from HWMStore. This may be changed in next PRs.

## Checklist

* [X] Commit message and PR title is comprehensive
* [ ] Keep the change as small as possible
* [X] Unit and integration tests for the changes exist
* [X] Tests pass on CI and coverage does not decrease
* [X] Documentation reflects the changes where applicable
* [X] `docs/changelog/next_release/<pull request or issue id>.<change type>.rst` file added describing change
  (see [CONTRIBUTING.rst](https://github.com/MobileTeleSystems/onetl/blob/develop/CONTRIBUTING.rst) for details.)
* [ ] My PR is ready to review.

### CODE REVIEW

File: onetl/strategy/hwm_strategy.py [hunk 1]
```diff
@@ -57,28 +54,70 @@ def enter_hook(self) -> None:
...
             # entering strategy context, HWM will be set later by DBReader.run or FileDownloader.run
-            log.debug("|%s| HWM will not be loaded, skipping", class_name)
+            log.debug("|%s| HWM will not be fetched, skipping", class_name)
+            return
+
+        hwm_store = HWMStoreStackManager.get_current()
+        log.info("|%s| Fetching HWM from %s:", class_name, hwm_store.__class__.__name__)
+        log_with_indent(log, "name = %r", self.hwm.qualified_name)
+
+        result = hwm_store.get_hwm(self.hwm.qualified_name)
```
Comment (maxim-lixakov) [inline onetl/strategy/hwm_strategy.py:66]:
```suggestion
        log_with_indent(log, "name = %r", self.hwm.name)

        result = hwm_store.get_hwm(self.hwm.name)
```
у нас в новых hwm же нет уже `qualified_name`?

Comment (dolfinus) [reply onetl/strategy/hwm_strategy.py:66]: Я в отдельном PR буду обновлять etl-entities до 2.1, там и атрибуты переименую. А то и так уже много изменений

File: tests/tests_integration/tests_strategy_integration/test_strategy_incremental_batch.py [hunk 2]
```diff
@@ -284,30 +241,34 @@ def test_postgres_strategy_incremental_batch_different_hwm_type_in_store(
         spark=spark,
     )
 
+    hwm_name = secrets.token_hex(5)
+
     reader = DBReader(
         connection=postgres,
         source=load_table_data.full_name,
-        hwm=DBReader.AutoDetectHWM(name=secrets.token_hex(5), column=hwm_column),
+        hwm=DBReader.AutoDetectHWM(name=hwm_name, column="hwm_int", expression="hwm_int"),
```
Comment (maxim-lixakov) [inline tests/tests_integration/tests_strategy_integration/test_strategy_incremental_batch.py:249]: мы в `DBReader.AutoDetectHWM` тоже ведь переводим `column` -> `source`  и добавляем   `expression` ?

Comment (dolfinus) [reply tests/tests_integration/tests_strategy_integration/test_strategy_incremental_batch.py:249]: Да, но я решил сделать это в другом PR, а в этом оставить etl-entities 2.0 со старыми именами

File: onetl/strategy/hwm_strategy.py [hunk 3]
```diff
@@ -57,28 +54,70 @@ def enter_hook(self) -> None:
...
+                f"""
+                HWM type {type(other_hwm).__name__!r} fetched from HWM store
+                does not match current HWM type {hwm_type.__name__!r}.
+
+                Please:
+                * Check that you set correct HWM name, it should be unique.
+                * Check that your HWM store contains valid value and type for this HWM name.
+                """,
+            )
+            raise TypeError(message)
```
Comment (maxim-lixakov) [inline onetl/strategy/hwm_strategy.py:98]: мы в `DBReader`  выбрасываем `RuntimeError` в случаях  неправильного использования hwm, может раз мы тут тоже сообщаем пользователю инструкцию по диагностики то тоже бросать `RuntimeError`?
ну и как то стандартизировать, а то в тестах по похожим ошибкам вылетают разные исключения, например - `test_greenplum_strategy_incremental_wrong_wm_type`

File: tests/tests_integration/tests_strategy_integration/tests_incremental_strategy_integration/test_strategy_increment_greenplum.py [hunk 4]
```diff
@@ -117,8 +114,9 @@ def test_greenplum_strategy_incremental(
...
     "hwm_column, exception_type, error_message",
     [
-        ("float_value", ValueError, "value is not a valid integer"),
-        ("text_string", KeyError, "Unknown HWM type 'string'"),
+        ("float_value", ValueError, "Expression 'float_value' returned values"),
+        ("text_string", RuntimeError, "Cannot detect HWM type for"),
+        ("unknown_column", Exception, "column .* does not exist"),
     ],
 )
 def test_greenplum_strategy_incremental_wrong_wm_type(
```
Comment (maxim-lixakov) [inline tests/tests_integration/tests_strategy_integration/tests_incremental_strategy_integration/test_strategy_increment_greenplum.py:122]:
```suggestion
def test_greenplum_strategy_incremental_wrong_hwm_type(
```

File: onetl/strategy/hwm_strategy.py (hunk 3 above)
Comment (dolfinus) [reply onetl/strategy/hwm_strategy.py:98]: Да, я немного думал в этом направлении. Нужно будет завести в модуле onetl.exception разные типы исключений под кейсы с валидацией, недоступностью источника и т.п.

Конкретно тут логика была в том, что у нас получились два значения несовместимых между собой типов, и тут логично выбрасывать TypeError

File: onetl/db/db_reader/db_reader.py [hunk 5]
```diff
@@ -600,115 +520,238 @@ def run(self) -> DataFrame:
...
+            self._prepare_hwm(strategy, self.hwm)
+
+        elif isinstance(strategy, HWMStrategy):
+            raise RuntimeError(f"{strategy_name} cannot be used without {class_name}(hwm=...)")
+
+    def _prepare_hwm(self, strategy: HWMStrategy, hwm: ColumnHWM):
+        if not strategy.hwm:
+            # first run, HWM was never uploaded to from HWM Store
+            strategy.hwm = self._autodetect_hwm(hwm)
+            strategy.fetch_hwm()
```
Comment (maxim-lixakov) [inline onetl/db/db_reader/db_reader.py:560]: мы теперь всегда получаем hwm из `hwm_store`, даже если пользователь сам передал `hwm=ColumnIntHWM(...)` ?

Comment (dolfinus) [reply onetl/db/db_reader/db_reader.py:560]: Нам в любом случае нужно получать hwm.value из HWMStore, иначе в самом HWMStore не будет никакого смысла.

А вот насчет того, что теперь автоопределение типа HWM используется всегда - тут надо подумать, возможно стоит вернуть как было

Comment (dolfinus) [reply onetl/db/db_reader/db_reader.py:560]: Тут такой момент, если пользователь передал `DBReader(hwm=AutoDetectHWM(value=123, name=...))` или `DBReader(hwm=ColumnIntHWM(value=123, name=...))`, и мы будем смотреть исключительно на значение, переданное пользователем, то все стратегии превращаются в тыкву, потому что будут читать только то, что пользователь передал в сам класс, а HWM store будет полностью игнорироваться

Comment (maxim-lixakov) [reply onetl/db/db_reader/db_reader.py:560]: справедливо

**Pull request merged**